intermediate_source/rpc_async_execution.rst

```diff
@@ -0,0 +1,504 @@
+callee to execute the function in that request until that function returns.
+This works for many use cases, but there is one caveat. If the user function
+blocks on IO, e.g., with nested RPC invocation, or signaling, e.g., waiting for
+a different RPC request to unblock, the RPC thread on the callee will have to
+idle waiting until the IO finishes or the signaling event occurs. As a result,
+RPC callees are likely to use more threads than necessary. The cause of this
+problem is that RPC treats user functions as black boxes, and knows very little
+about what happens in the function. To allow user functions to yield and free
+RPC threads, more hints need to be provided to the RPC system.
+
+Since v1.6.0, PyTorch addresses this problem by introducing two new concepts:
+
+* A `torch.futures.Future <https://pytorch.org/docs/master/futures.html>`__ type
+  that encapsulates an asynchronous execution, which also supports installing
+  callback functions.
+* An `@rpc.functions.async_execution <https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.functions.async_execution>`__
+  decorator that allows applications to tell the callee that the target function
+  will return a future and can pause and yield multiple times during execution.
+
+With these two tools, the application code can break a user function into
+multiple smaller functions, chain them together as callbacks on ``Future``
+objects, and return the ``Future`` that contains the final result. On the callee
+side, when getting the ``Future`` object, it installs subsequent RPC response
+preparation and communication as callbacks as well, which will be triggered
+when the final result is ready. In this way, the callee no longer needs to block
+one thread and wait until the final return value is ready Please refer to the
+API doc of
+`@rpc.functions.async_execution <https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.functions.async_execution>`__
+for simple examples.
+
+Besides reducing the number of idle threads on the callee, these tools also help
+to make batch RPC processing easier and faster. The following two sections of
+this tutorial demonstrate how to build distributed batch-updating parameter
+server and batch-processing reinforcement learning applications using the
+`@rpc.functions.async_execution <https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.functions.async_execution>`__
+decorator.
+
+Batch-Updating Parameter Server
+-------------------------------
+
+Consider a synchronized parameter server training application with one parameter
+server (PS) and multiple trainers. In this application, the PS holds the
+parameters and waits for all trainers to report gradients. In every iteration,
+it waits until receiving gradients from all trainers and then updates all
+parameters in one shot. The code below shows the implementation of the PS class.
+The ``update_and_fetch_model`` method is decorated using
```

Comment:
nit: It might be good for users to provide some more explanation about how `update_and_fetch_model` works (maybe comments in the function explaining certain parts)